jdeeco-core/src/cz/cuni/mff/d3s/deeco/runtime/DEECo.java

```diff
@@ -0,0 +1,226 @@
+		
+		runtime.start();
+		running = true;
+	}
+
+	public void stop() throws InvalidOperationException {
+		if(!running)
+			throw new InvalidOperationException("stop");		
+		
+		runtime.stop();
+		running = false;
+	}
+	
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T extends DEECoPlugin> T getPluginInstance(Class<T> pluginClass) {
+		return (T) pluginsMap.get(pluginClass);
+	}
+	
+	@Override
+	public RuntimeFramework getRuntimeFramework()
+	{
+		return runtime;
+	}
+	
+	@Override
+	public AnnotationProcessor getProcessor() {
+		return processor;
+	}	
+	
+	@Override
+	public RuntimeMetadata getRuntimeMetadata() {
+		return model;
+	}
+	
+	@Override
+	public boolean isRunning()
+	{
+		return running;
+	}
+	
+	class DependencyNode {
+		DEECoPlugin plugin; 
+		List<DependencyNode> dependantPlugins = new ArrayList<>();
+		int dependencyCount;
+		
+		DependencyNode(DEECoPlugin plugin) {
+			this.plugin = plugin;
+			this.dependencyCount = plugin.getDependencies().size();
+		}
+		
+		DependencyNode() {
+			dependencyCount = 0;
+		}
+	}
+	
+	class DependencyNodeComparator implements Comparator<DependencyNode>
+	{
+		@Override
+		public int compare(DependencyNode o1, DependencyNode o2) {			
+			return Integer.compare(o1.dependencyCount, o2.dependencyCount);
+		}		
+	}
+	
+	@SuppressWarnings("rawtypes")
+	List<DependencyNode> constructDependencyNodes(DEECoPlugin[] plugins) throws PluginDependencyException {		
+		List<DependencyNode> dependencyNodes = new ArrayList<>();		
+		Map<Class, DependencyNode> knownPlugins = new HashMap<>();
+		
+		for (DEECoPlugin p: plugins) {
+			DependencyNode node = new DependencyNode(p);
+			
+			knownPlugins.put(p.getClass(),node);
+			dependencyNodes.add(node);
+		}
+		
+		for (DependencyNode node: dependencyNodes) {
+			for (Class<? extends DEECoPlugin> pluginClass : node.plugin.getDependencies()) {					
+				if (knownPlugins.containsKey(pluginClass)) {
+					knownPlugins.get(pluginClass).dependantPlugins.add(node);						
+				} else {
+					throw new MissingDependencyException(node.plugin.getClass(), pluginClass);
+				}
+			}
+		}	
+		
+		return dependencyNodes;
+	}		
+
+	void initializePlugins(DEECoPlugin[] plugins) throws PluginDependencyException {
+		List<DependencyNode> nodes = constructDependencyNodes(plugins);
+		Queue<DependencyNode> queue = new PriorityQueue<DEECo.DependencyNode>(new DependencyNodeComparator());
+		
+		for(DependencyNode n : nodes)
+		{
+			queue.add(n);
+		}
+
+		while (!queue.isEmpty()) {
+			DependencyNode n = queue.remove();				
+
+			if (n.dependencyCount == 0) {
+				n.plugin.init(this);
+				pluginsMap.put(n.plugin.getClass(), n.plugin);				
+
+				for (DependencyNode dependantPlugin : n.dependantPlugins) {
+					queue.remove(dependantPlugin);
+					dependantPlugin.dependencyCount--;
+					queue.add(dependantPlugin);					
+				}
+				
+			} else {
+				throw new CycleDetectedException();
+			}
+		}	
+	}
+	
+	private void createRuntime() {
```

Comment:
This is generally fine. But I would not call the RuntimeFramework a plugin. If it was a plugin, then it would be added to plugins and initialized with the rest of plugins. Actually the content of this method should be part of the initialization code of the plugin.